### PR TITLE
HTML Stmt IR with conceptual code and device code.

### PR DESCRIPTION
**Pull request by mcourteaux**
### Features and functionality in this PR

 - Implemented the Generator options to emit `conceptual_stmt` and `conceptual_stmt_html`:
image

   The Conceptual Stmt HTML is particularly useful when working on GPU schedules, as you get to see the Stmt code just before offloading the GPU-related loops to LLVM and getting device code back. This addresses #7519.

 - Added "Jump to Device code" buttons (above in green).
 - Added a Device Code pane to the page, which replaces the inline buffer in the Stmt IR. Buffers' content are no longer included in the Stmt IR.
 - Collapsing the subtrees does no longer recompute the costs (movement and computation) of the nodes. Data moving and computation costs are always shown both for the line itself, and for the whole subtree (left is for a the subtree (aka: including children), and right is the line itself):
image
Note that I added one extra color in the coloring scheme for zero-cost nodes: transparent.
 - Added some shading of GPU-related for-loops and functions. I tried to make it subtle but clear when you actually want to see the background shading.
 - Added `device_code` emit option for Generators, which will produce a `.device_code` file. This currently only works if the device code is generated by `OffloadGPULoops`. I'm not familiar with Hexagon.

### HTML-CSS-JS specific changes:

In general, I would finally consider the clean (no-VizTree) HTML Stmt IR as sane. I'm not a front-end developer, but I know a few things of keeping things simple.

 - ~~Split the code (HTML/CSS/JS) from the VizIR stuff into the parts that belong to the clean Stmt IR HTML and the parts that belong to the Viz Tree stuff. These are now separate files and are included in the generated result depending on whether the VizTree is enabled.~~
 - ~~Added the possibility to set `HL_HTML_VIZTREE=1` as an environment variable which will enable the VizTree generation.~~
 - **UPDATE**: Nuked all the VizTree code. Additionally renamed the word "tabs" to "panes" in the code.
 - Got rid of dependencies in the clean Stmt IR code. No more bootstrap, no more bootstrap icons, no more jQuery, no more flextrees, or whatever. This addresses the concerns raised in #7789.
 - Improved the performance of resizing the splits by showing a split previewer instead of live-dragging the panes.
 - Panes collapsing logic is taken out of the dragging behaviour. Collapsing is now done by clicking the buttons on the resize bars.
 - Added the ability to jump to device code from certain loops. This is rather hacky and depends again on LLVM generating comments based on the BasicBlock name.
 - Added clear distinctions between "Jump to Host code" (Red) and "Jump to Device code" (Red) buttons.
 - Collapsing of subtrees in the Stmt IR is implemented with checkboxes and corresponding CSS rules that act on siblings on `:checked` checkbox inputs.
 - Bugfixed some invalid HTML constructs.

### Relevant other changes:

 - Now print `vector_reduce(op, data)` now as `vector_reduce_op(data)`. This turns `vector_reduce(0, data)` into `vector_reduce_add(data)`, which is better to read. The reason this VectorReduce::Operator printed as an integer is because there is no such thing as a `IRVisitor::visit(VectorReduce::Operator)`.
 - Refactored a bit how the Generator processes the emit options: `HL_EXTRA_OUTPUTS` would cause the default emit-options to be not used when originally the Generator was invoked without any explicit emit flags. No emit flags means to generate "static_library,cpp_registration,c_header", but once you specify `HL_EXTRA_OUTPUTS` that would cause the logic skip installing the default ones. @abadams @steven-johnson Please verify that this behavior is fine. I was using this `HL_EXTRA_OUTPUTS=stmt_html,conceptual_stmt_html make` in some of the `apps/`, but would fail because now the `c_header` and the `static_library` options were no longer implied.

### PR status

 - ~~Currently, the VizTree stuff is not revived yet, as my focus was to clean up the messy HTML and JavaScript code for the clean Stmt IR HTML. I didn't throw out anything fundamental yet from the VizTree. Opinions on this are welcome! Of course it will be easier to delete the remaining stuff than to revive it, and honestly, I would rather delete it than revive as I personally have absolutely no use for it.~~
 **UPDATE:** I nuked the VizTree code. This brings the PR to be ready for review I believe.

 - Tested the result (of the clean Stmt IR HTML) on Chromium and Firefox.

 - Tested with:
   - [x] CPU-only (x64)
   - [x] CUDA kernels (PTX)
   - [ ] OpenCL
   - [ ] Metal
   - [x] OpenGL Compute Shaders
   - [ ] Hexagon
   - [ ] Direct3D12
   - [ ] Vulkan

   Only PTX code right now is syntax highlighted, and the other should just be pasted in without any coloring. _Syntax highlighting for other languages welcome!_

### Known issues

 - ~~The left collapser icon doesn't change in Firefox, as Firefox has no support for `:has(+ sibling)` CSS selectors. This could be replaced with JavaScript code, but I like to avoid JS, and the visual impact of this issue is minor.~~
 - Hexagon code is currently still generated immediately as compiled binary, without an assembly output, so we can't currently include that code. I pesonally have never used it, so I'm rather clueless on what is desirable, nor do I know how to test it.
 - Jump-to-device-code buttons are only working with LLVM CodeGen backends.
 - Vulkan code is not human-readable. Maybe inserting the LLVM IR is a useful intermediate step.

---

@steven-johnson @antonysigma @maaz139 @abadams were all involved in this discussion at some point.

### CODE REVIEW

**Comment by steven-johnson** [comment]
I'm not sure that I'm the best reviewer for this -- tagging @abadams @zvookin @vksnk to see if they would like to fill in for me

**Comment by mcourteaux** [comment]
@steven-johnson I believe you work at Google? Does anybody you know use the VizTree stuff? I highly doubt it. @abadams You work at Adobe I think? Does anybody you know use the VizTree stuff? Do you have opinions on whether to keep the VizTree thing or not? It looks pretty, but I don't see any use for it.

**Comment by abadams** [comment]
I'm not aware of anyone using the VizTree stuff here at Adobe. You're the person most invested in the stmt html output (though that may change with the functionality you're adding!), so I defer to you entirely on what should and should not be included.

**Comment by mcourteaux** [comment]
> I'm not aware of anyone using the VizTree stuff here at Adobe. You're the person most invested in the stmt html output (though that may change with the functionality you're adding!), so I defer to you entirely on what should and should not be included.

Oh wow? I'm surprised. Nobody uses the generated HTML? How do people review the schedule they built for a certain pipeline? You can't just go based off the `print_loop_nest()` or timing information, right? This Stmt HTML is what gets me to consistently to improve all pipelines I work on to get 10x faster than a simple but blind tile-parallel schedule with some obvious `.compute_at()` improvements, sprinkled with `.vectorize()`.

Almost 100% of the time, my workflow is iterating between reviewing the stmt code looking for inefficiencies, and improving the schedule, and confirming the improvement by benchmarks.

**Comment by vksnk** [comment]
> This Stmt HTML is what gets me to consistently to improve all pipelines I work on to get 10x faster than a simple but blind tile-parallel schedule with some obvious .compute_at() improvements, sprinkled with .vectorize().

> Almost 100% of the time, my workflow is iterating between reviewing the stmt code looking for inefficiencies, and improving the schedule, and confirming the improvement by benchmarks.

Just personal experience: I 100% agree that it's far better to iterate on the schedule by looking for inefficiencies in the generated statement is. That being said, I pretty much always just use a plain statement in terminal/text editor instead of the HTML one. Previously, I felt that overhead of switching to browser to look at the statement was too high for me. I should try a new visualizer though, it looks pretty interesting.

I don't remember if advertised new visualizer for internal users at Google, but we certainly can once it stabilizes (maybe it's already stable, but there were a lot of changes recently, so I am not sure).

**Comment by abadams** [comment]
I also use the text-based stmt and assembly.

**Comment by mcourteaux** [comment]
Okay, then I would just discard this feature, and maybe park it in a stale branch, if anybody would like to revive it at one point. I'll finalize this PR tonight or tomorrow.

**Comment by mcourteaux** [comment]
> I pretty much always just use a plain statement in terminal/text editor instead of the HTML one.

While I work 95% of the time in the terminal for almost all I do, I think I'd prefer the HTML one for folding, syntax highlighting, and variable cross-matching, parenthesis highlighting, and the jump-to-assembly buttons. Do you have any syntax highlighting when investigating in the terminal? A vim syntax file? Some rainbow-parenthesis plugins? Or just "black and white"?

**Comment by abadams** [comment]
Just black and white. I think with the ability to hop between asm and stmt easily I may start using the html version though.

**Comment by steven-johnson** [comment]
The HTML version would probably be preferable now; I think it's mainly inertia that keeps the core devs on the .stmt file, with a smattering of a slightly less convenient workflow in some cases (e.g.: I can open the .stmt file in an editor that will just auto-refresh every time I re-run the Generator, whereas .html requires a manual refresh. Trivial? Yes, but these are the things that add up to inertia.)

File: src/irvisualizer/html_template_StmtToHTML.js [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+        top: lno * line_height
+    });
+}
+
+function scrollToDeviceCode(lno) {
+    var device_code_tab = document.getElementById("device-code-tab");
+    const lineSpans = device_code_tab.querySelectorAll('span.line');
+    line = lineSpans[lno - 1]
+    line.scrollIntoView({
+        behavior: "smooth",
```
Comment (antonysigma) [inline src/irvisualizer/html_template_StmtToHTML.js:149]: My linter catches some errors; all are minute. I am attaching the log here for your reference.

```javascript
/home/antony/Projects/halide/src/irvisualizer/html_template_StmtToHTML.js
   33:10  error  'scrollToCode' is defined but never used        no-unused-vars
   36:5   error  'makeCodeVisible' is not defined                no-undef
   57:13  error  Unexpected console statement                    no-console
   61:17  error  'rect' is defined but never used                no-unused-vars
   71:39  error  'event' is defined but never used               no-unused-vars
   85:9   error  Unexpected console statement                    no-console
  119:10  error  'collapse_tab' is defined but never used        no-unused-vars
  129:10  error  'scrollToHostAsm' is defined but never used     no-unused-vars
  137:10  error  'scrollToDeviceCode' is defined but never used  no-unused-vars
  140:5   error  'line' is not defined                           no-undef
  141:5   error  'line' is not defined                           no-undef
  142:28  error  Unexpected trailing comma                       comma-dangle
```

Comment (antonysigma) [reply src/irvisualizer/html_template_StmtToHTML.js:149]: This is the content of my `.eslint.js` config:

```javascript
module.exports = {
  env: {
    es6: true,
    node: true,
    browser: true
  },
  extends: [
    'eslint:recommended',
  ],
  parserOptions: {
    // Enable JSX support
    ecmaFeatures: { jsx: true },
    ecmaVersion: 2018,
    sourceType: 'module'
  },
  //parser: 'babel-eslint',
  rules: {
    // Add any additional rules here
    'react/prop-types': 0
  }
}
```

File: src/irvisualizer/html_template_StmtToViz_dependencies.html [hunk 2]
```diff
@@ -0,0 +1,4 @@
+<!-- Hierarchy links -->
+<link rel='stylesheet' href='https://unpkg.com/treeflex/dist/css/treeflex.css'>
+
+<script src='http://code.jquery.com/jquery-1.10.2.js'></script>
```
Comment (antonysigma) [inline src/irvisualizer/html_template_StmtToViz_dependencies.html:4]: Please upgrade to JQuery 3.7.x in order to resolve CVEs. I can help flush out deprecated JQuery API calls after that.

**Review by antonysigma — CHANGES_REQUESTED** [review]
Many thanks @mcourteaux for the major revamp! I haven't tested the code yet. In the meantime, I quickly ran `eslint` liner tool to flush out Javascript errors. I am attaching the logs below for your reference.

Regards,
Antony

File: src/irvisualizer/html_template_StmtToViz_dependencies.html (hunk 2 above)
Comment (mcourteaux) [reply src/irvisualizer/html_template_StmtToViz_dependencies.html:4]: I will nuke StmtToViz. I started by separating the two first cleanly to work on the StmtHTML. Thanks for the reminder tho!

File: src/irvisualizer/html_template_StmtToHTML.js (hunk 1 above)
Comment (mcourteaux) [reply src/irvisualizer/html_template_StmtToHTML.js:149]: Thank you for these! I will look into it. Is there ways to suppress some of these? Some of these "unused functions" are of course not unused, as the HTML is sprinkled with `onclick="scrollToHostAsm()"` for example.

Comment (mcourteaux) [reply src/irvisualizer/html_template_StmtToHTML.js:149]: Yeah, I found that this should work:

```js
function scrollToDeviceCode(lno) {  // eslint-disable-line no-unused-vars
```

File: src/irvisualizer/html_template_StmtToHTML.js [hunk 3]
```diff
@@ -0,0 +1,151 @@
+// Highlighting 'matched' elements in IR code
+function addHighlight(selector) {
+    var matchedElements = document.querySelectorAll(selector + ' .matched');
+
+    matchedElements.forEach(function(element) {
+        element.addEventListener('mouseover', function() {
+            var idPrefix = this.id.split('-')[0];
+            var matchingElements = document.querySelectorAll(selector + ' .matched[id^="' + idPrefix + '-"]');
+
```
[the remark this answers is not in the document]
Comment (mcourteaux) [reply src/irvisualizer/html_template_StmtToHTML.js:9]: Nice! Great suggestion!

**Comment by mcourteaux** [comment]
@antonysigma I attempted to fix your immediate feedback. I haven't looked at running eslint myself, but if you'd like to give it another go, feel free to! I addressed them I believe.

File: src/irvisualizer/html_template_StmtToHTML.js (hunk 1 above)
Comment (antonysigma) [reply src/irvisualizer/html_template_StmtToHTML.js]: > I haven't looked at running eslint myself, but if you'd like to give it another go, feel free to!

For sure. I am linting the `StmtToHTML.js` only. 

```javascript
halide/src/irvisualizer/html_template_StmtToHTML.js
   34:5   error  'makeCodeVisible' is not defined   no-undef
   69:39  error  'event' is defined but never used  no-unused-vars
  146:51  error  'device_code_Tab' is not defined   no-undef
```

@mcourteaux Ping me again if you want a sanity check for legacy web browsers:

image

File: src/Module.cpp [hunk 4]
```diff
@@ -49,7 +49,10 @@ std::map<OutputFileType, const OutputInfo> get_output_info(const Target &target)
         {OutputFileType::schedule, {"schedule", ".schedule.h", IsSingle}},
         {OutputFileType::static_library, {"static_library", is_windows_coff ? ".lib" : ".a", IsSingle}},
         {OutputFileType::stmt, {"stmt", ".stmt", IsMulti}},
+        {OutputFileType::conceptual_stmt, {"coneptual_stmt", ".conceptual.stmt", IsMulti}},
```
Comment (antonysigma) [inline src/Module.cpp:52]: Is it a typo? `coneptual -> conceptual`?

```
Unhandled exception: Error: Unrecognized emit option: conceptual_stmt is not one of [assembly bitcode c_header c_source compiler_log cpp_stub featurization function_info_header hlpipe llvm_assembly object python_extension pytorch_wrapper registration schedule static_library stmt coneptual_stmt stmt_html conceptual_stmt_html device_code], ignoring.
```

**Review by antonysigma — COMMENTED** [review]
@mcourteaux 

I have a few minor UX related questions. Please read the inline comments.

Regards,
Antony

Update: I like the green arrows to hop between the NVPTX panel and the Stmt panel. Not all offloaded GPU code has the green arrow, but yeah it exceed all my expectations. Attaching a demo HTML here to reach a broader audience, including but not limited to @abadams and @vksnk .

image

[demo.stmt.html.gz](https://github.com/halide/Halide/files/12612579/demo.stmt.html.gz)

File: src/StmtToViz.cpp [hunk 5]
```diff
@@ -2851,123 +3043,162 @@ class IRVisualizer {
...
+    // Generate a resizing bar to control the width of code and visualization panes
+    void generate_resize_bar(int num) {
+        stream << "<div class='resize-bar' id='resize-bar-" << num << "'>\n";
+        stream << "   <div class='collapse-btns'>\n";
+        stream << "     <div>\n";
+        stream << "      <button class='collapse-left' onclick='collapseTab(" << num << ")' title='Collapse pane on the left'></span>\n";
+        stream << "       </button>\n";
+        stream << "     </div>\n";
+        stream << "     <div>\n";
+        stream << "       <button class='collapse-right' onclick='collapseTab(" << (num + 1) << ")' title='Collapse pane on the right'></span>\n";
```
[the remark this answers is not in the document]
Comment (mcourteaux) [reply src/StmtToViz.cpp:3189]: Yeah, I was trying to help distinguish this by shading the button. My main concern is screen estate. The green seperator is already quite thick, and wasting screen estate. For now, I decided to not waste more space just to be able to put a bigger arrow icon. I never really looked at the icons anyway. I just picked a few unicode arrows, slammed them on, and realized I was looking at the color of the button instead of the icon.

[the remark this answers is not in the document]
Comment (mcourteaux) [reply src/StmtToViz.cpp:3189]: > A minor UX issue: Once I clicked the top button to collapse the stmt tab, the bottom button clears out the assembly tab.

Wait, that's exactly what it's supposed to do right? The button toggles the collapsedness of the tab left or right of it. Top button controls the tab on the left. Bottom button controls the tab on the right. You just click them to toggle whether they are collapsed or not.

File: src/Module.cpp (hunk 4 above)
Comment (mcourteaux) [reply src/Module.cpp:52]: Thanks for spotting this. My light form of dyslexia makes spotting these hard.

File: src/irvisualizer/html_template_StmtToHTML.js (hunk 1 above)
Comment (mcourteaux) [reply src/irvisualizer/html_template_StmtToHTML.js]: I don't know how to interpret the screenshot of this legacy browser check. There are a bunch of numbers, but I don't know what they all indicate. Please keep in mind that this Halide pipeline visualizer HTML is a tool for developers who work on a desktop. The browsers we care about are all desktop based: Firefox, Chrome, Edge, Safari, Opera, would probably cover most of them.

**Comment by mcourteaux** [comment]
> Attaching a demo HTML here to reach a broader audience, including but not limited to @abadams and @vksnk .

That demo file doesn't work, as the dependent files (the `.js` and `.css`) are not inlined, and instead linked to the fixed location on your machine. I made the inlining an option to increase development iteration speed. Inlining them, and therefore making them self-contained, can be done here:

https://github.com/halide/Halide/blob/de1970f2d63542cc5b49020fc3e1b1472f932498/src/StmtToViz.cpp#L20-L25

File: src/StmtToViz.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (antonysigma) [reply src/StmtToViz.cpp:3189]: Marked as resolved.

**Review by antonysigma — APPROVED** [review]

**Comment by mcourteaux** [comment]
New demo file: [bgu_html_demo.zip](https://github.com/halide/Halide/files/12629381/bgu_html_demo.zip)

**Comment by mcourteaux** [comment]
Okay, status update:

 - I want to get this PR out my head.
 - The usability is not perfect, because of the known issues (above in the OP), but it is great for CPU-only and GPU-CUDA code.
 - The additional features are probably due for someone that actively works with that specific codegen setup. I for example have never used Hexagon, nor Vulkan, nor Direct3D, and I'm generally speaking clueless and not motivated enough to start figuring out how to make the `jump-to-device-code` buttons work for all these languages. Same thing with the syntax highlighting and code block folding. It is possible, and deseriable, but I'm not invested enough in these other languages right now to make it work.

Bottom line: I think this PR is ready to be reviewed and merged. Yet there is some unexploited potential for other GPU devices.

**Comment by mcourteaux** [comment]
It's been quiet for a while. @steven-johnson can you coordinate from here? PR is ready. 😃

File: apps/bilateral_grid/bilateral_grid_generator.cpp [hunk 6]
```diff
@@ -84,7 +84,7 @@ class BilateralGrid : public Halide::Generator<BilateralGrid> {
             // nothing
         } else if (get_target().has_gpu_feature()) {
             // 0.50ms on an RTX 2060
-
+            std::printf("GPU SCHEDULING\n");
```
Comment (abadams) [inline apps/bilateral_grid/bilateral_grid_generator.cpp:87]: Debugging code left in

File: src/Lower.cpp [hunk 7]
```diff
@@ -432,6 +432,9 @@ void lower_impl(const vector<Function> &output_funcs,
         }
     }
 
+    // Make a copy of the Stmt code, before we lower anything to 2GL.
```
Comment (abadams) [inline src/Lower.cpp:435]: I don't know what "2GL" means

File: src/Module.cpp [hunk 8]
```diff
@@ -325,6 +328,12 @@ struct ModuleContents {
     MetadataNameMap metadata_name_map;
     bool any_strict_float{false};
     std::unique_ptr<AutoSchedulerResults> auto_scheduler_results;
+
+    /** This is a copy of the code throughout the lowering process, which
+     * reflects best the actual pipeline, without introducing assembly generated
+     * from device-specific offloads (such as Cuda PTX). In other words, we'd
+     * like to keep this 3GL. */
```
Comment (abadams) [inline src/Module.cpp:335]: I don't know what 3GL means

File: src/Module.h [hunk 9]
```diff
@@ -164,6 +167,17 @@ class Module {
     const std::vector<Module> &submodules() const;
     // @}
 
+    /** Tries to locate the offloaded CUDA PTX assembly contained in this Module.
+     * Might return a nullptr in case such buffer is not present in this Module.
+     */
+    const Buffer<> *get_cuda_ptx_assembly_buffer() const;
```
Comment (abadams) [inline src/Module.h:173]: Buffers are already a pointer class that can be checked for nullity with buffer.defined(), so it's probably better to return an buffer than a pointer to a buffer.

File: src/Module.h [hunk 10]
```diff
@@ -201,6 +215,13 @@ class Module {
 
     /** Set whether this module uses strict floating-point directives anywhere. */
     void set_any_strict_float(bool any_strict_float);
+
+    /** Remember the Stmt during lowing which is still 3GL, before device-specific
```
Comment (abadams) [inline src/Module.h:219]: Same comment here about not knowing what 3GL means.

File: src/StmtToHTML.cpp [hunk 11]
```diff
@@ -1560,6 +1645,28 @@ class HTMLCodePrinter : public IRVisitor {
         scope.pop(op->name);
     }
 
+    std::string ForType_to_string(ForType type) {
```
Comment (abadams) [inline src/StmtToHTML.cpp:1648]: This should probably be: 
```
std::ostringstream oss; 
oss << type;
return oss.str();
```

To reuse the code already in IRPrinter.h/cpp

File: src/StmtToHTML.cpp [hunk 12]
```diff
@@ -2099,7 +2162,33 @@ class HTMLCodePrinter : public IRVisitor {
         print_text("(");
         print_type(op->type);
         print_text(")");
-        print_function_call("vector_reduce", {op->op, op->value});
+        std::string op_str;
+        switch (op->op) {
```
Comment (abadams) [inline src/StmtToHTML.cpp:2166]: This should use the operator<< already in IRPrinter.h (and the implementation in IRPrinter.cpp can be changed to be snake case if you like).

File: src/IRPrinter.cpp [hunk 13]
```diff
@@ -1085,10 +1085,35 @@ void IRPrinter::visit(const Shuffle *op) {
 }
 
 void IRPrinter::visit(const VectorReduce *op) {
+    std::string op_str;
```
Comment (abadams) [inline src/IRPrinter.cpp:1088]: This should use the existing operator<< for vector reduce op types (with the implementation changed to snake case if you like)

File: src/runtime/hexagon_dma.cpp [hunk 14]
```diff
@@ -710,7 +710,7 @@ WEAK int halide_hexagon_dma_power_mode_voting(void *user_context, halide_hexagon
         val = ~PW_SVS;
         break;
     default:
-        error(user_context) << "Hexagon: halide_hexagon_dma_power_voting power mode (", cornercase, ") not found\n";
+        error(user_context) << "Hexagon: halide_hexagon_dma_power_voting power mode (" << cornercase << ") not found\n";
```
Comment (abadams) [inline src/runtime/hexagon_dma.cpp:713]: Nice catch. How did you find it?

**Comment by abadams** [comment]
Generally speaking, looks good! I just had some minor comments.

**Comment by steven-johnson** [comment]
> It's been quiet for a while. @steven-johnson can you coordinate from here? PR is ready. 😃

For unrelated reasons, I'm not going to have the bandwidth to do the review and coordination on this for a while -- looks like @abadams is taking over (thanks!)

File: src/runtime/hexagon_dma.cpp (hunk 14 above)
Comment (mcourteaux) [reply src/runtime/hexagon_dma.cpp:713]: Good question. I only realize now that you'd be very interested in catching this automatically. One of the compilers I used over the last weeks was complaining about it. Could have been clamg on macOS.

**Comment by mcourteaux** [comment]
I'm currently on holidays. I'll address these comments properly when I'm back. The terms 2GL and 3GL were introduced into this discussion by @antonysigma in the original issue. The [wiki page](https://en.m.wikipedia.org/wiki/Programming_language_generations) describes it quite well. I am willing to change it of course, I also didn't know the terms.

**Comment by mcourteaux** [comment]
@steven-johnson If the build errors are not my fault, then @abadams this is ready.

File: src/IRPrinter.cpp (hunk 13 above)
Comment (abadams) [reply src/IRPrinter.cpp:1088]: I see you changed the implementation of operator<< to use snake case, so I think the changes to this function can be mostly reverted

**Comment by abadams** [comment]
One more super-minor comment where I think you missed one of the suggested changes. Also if you do a merge with main it should unbreak the builds.

**Comment by mcourteaux** [comment]
Oh well caught! Thanks. Merged with main as well. Let's see :)

**Comment by mcourteaux** [comment]
Doing some final experimenting with the colorscheme of the cost model. Hold off on merging for a few hours.

**Comment by mcourteaux** [comment]
Color palette of the cost model is now more interpretable: 

image

This is ready to merge, IMO. Did some final testing today.

**Comment by abadams** [comment]
Only failure is unrelated and will be fixed by #7868. Merging.

**Review by abadams — APPROVED** [review]
